### PR TITLE
Use ffmpeg-headless where applicable.

### PR DESCRIPTION
**Pull request by vs49688**
###### Description of changes
Following on from https://github.com/NixOS/nixpkgs/pull/196263. Now that we have `ffmpeg-headless`, use it!

Specifically, we can now cleanup `ffmpeg` override for `pipewire`.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by vs49688** [comment]
@Atemu from https://github.com/NixOS/nixpkgs/pull/196263#issuecomment-1279970187
> This would also be helpful for SDL bootstrap.

Where is this? I'd like to include it in this PR, but can't find it.

**Comment by Atemu** [comment]
@vs49688 I was misremembering. It was pipewire which depended on SDL2.

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -15739,12 +15739,8 @@ with pkgs;
...
   pipewire = callPackage ../development/libraries/pipewire {
     # ffmpeg depends on SDL2 which depends on pipewire by default.
-    # Break the cycle by disabling pipewire support in our ffmpeg.
-    ffmpeg = ffmpeg.override {
-      SDL2 = SDL2.override {
-        pipewireSupport = false;
-      };
-    };
+    # Break the cycle by depending on ffmpeg-headless.
+    ffmpeg = ffmpeg-headless;
```
Comment (Atemu) [inline pkgs/top-level/all-packages.nix:15743]: What does pipewire use ffmpeg for? Just want to make sure we're not giving it an ffmpeg that's missing critical functionality.

File: pkgs/top-level/all-packages.nix [hunk 2]
```diff
@@ -18578,7 +18574,9 @@ with pkgs;
     ffmpeg = ffmpeg_5;
   };
 
-  ffmpegthumbnailer = callPackage ../development/libraries/ffmpegthumbnailer { };
+  ffmpegthumbnailer = callPackage ../development/libraries/ffmpegthumbnailer {
+    ffmpeg = ffmpeg-headless;
+  };
```
Comment (Atemu) [inline pkgs/top-level/all-packages.nix:18579]: Nit: I typically prefer making the package itself explicitly depend on the package variant; tweaking its inputs.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (vs49688) [reply pkgs/top-level/all-packages.nix:15743]: It's only used in https://github.com/PipeWire/pipewire/blob/master/spa/plugins/ffmpeg/ffmpeg.c and friends, and just for encoding/decoding, so only `libavcodec`. The `-headless` changes shouldn't affect this (except _maybe_ the change to `libva-minimal`, but I seriously doubt that would hurt).

Comment (Atemu) [reply pkgs/top-level/all-packages.nix:15743]: Might be good to add that as a comment here.

File: pkgs/top-level/all-packages.nix (hunk 2 above)
Comment (vs49688) [reply pkgs/top-level/all-packages.nix]: Done. I've updated `navidrome` and `unpaper` too.

**Comment by vs49688** [comment]
@ofborg build openjdk17

**Comment by vs49688** [comment]
@ofborg build javaPackages.openjfx17

**Comment by SuperSandro2000** [comment]
How do we know that eg cyanrip is not broken after this?

File: pkgs/development/libraries/ffmpegthumbnailer/default.nix [hunk 3]
```diff
@@ -33,7 +33,7 @@ stdenv.mkDerivation rec {
         This thumbnailer was designed to be as fast and lightweight as possible.
         The only dependencies are ffmpeg and libpng.
     ";
-    platforms = platforms.linux;
+    platforms = platforms.all;
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/ffmpegthumbnailer/default.nix:36]:
```suggestion
    platforms = platforms.unix;
```
darwin+linux is unix for meta.platforms.

Comment (vs49688) [reply pkgs/development/libraries/ffmpegthumbnailer/default.nix:36]: Ah, cool, I will fix that up.

**Comment by vs49688** [comment]
> How do we know that eg cyanrip is not broken after this?

* I've manually tested `cyanrip`, `unpaper`, `navidrome`, and `ffmpegthumbnailer`.
* `openjfx*` only uses `libavcodec`, which isn't changed in the headless variant.
* I've tested `gst-libav` via `unoconv` -> `libreoffice` -> `gst-libav`.
    - Looking at the `gst-libav` source code shows that it only uses ffmpeg for some codecs and (de)muxers, nothing that would be really affected by the switch to headless.

I would test everything else more thoroughly, but my computer doesn't really have the chops to rebuild 5000+ packages.

**Comment by vs49688** [comment]
> > How do we know that eg cyanrip is not broken after this?
> 
>     * I've manually tested `cyanrip`, `unpaper`, `navidrome`, and `ffmpegthumbnailer`.
> 
>     * `openjfx*` only uses `libavcodec`, which isn't changed in the headless variant.
> 
>     * I've tested `gst-libav` via `unoconv` -> `libreoffice` -> `gst-libav`.
>       
>       * Looking at the `gst-libav` source code shows that it only uses ffmpeg for some codecs and (de)muxers, nothing that would be really affected by the switch to headless.
> 
> 
> I would test everything else more thoroughly, but my computer doesn't really have the chops to rebuild 5000+ packages.

Oh, and I haven't tested this, but Pipewire only uses libavcodec (via an SPA plugin), which isn't affected by the *-headless changes.

**Comment by SuperSandro2000** [comment]
> I would test everything else more thoroughly, but my computer doesn't really have the chops to rebuild 5000+ packages.

Thats why we target staging. If something breaks please follow up on it.

**Review by SuperSandro2000 — APPROVED** [review]

**Comment by vs49688** [comment]
ping on this?

**Pull request merged**